### DISCUSSION TITLE
Google auth invalid_grant

### DISCUSSION THREAD

**Discussion by pchasco** [pocketbase/pocketbase#442] [Q&A] [answered]
I am integrating PocketBase into a Flutter application and using Google auth. I am receiving the following error when attempting complete the authentication using the authViaOAuth2 method:

```json
{
  "errorDetails": "oauth2: cannot fetch token: 400 Bad Request\nResponse: {\n  \"error\": \"invalid_grant\",\n  \"error_description\": \"Bad Request\"\n}",
  "errorMessage": "Failed to authenticate."
}
```

Not much to go on. I can confirm that prior to calling the method, I am receiving the code value from Google, and I have configured Google auth in the console with the Client ID and Client Secret.

I have a demo project available on Github: https://github.com/pchasco/pocketbase_demo

To replicate, you'll need to configure web app credentials in the Google cloud console, and provide a synthetic redirect URI that you will use in the demo application. The demo application uses flutter_webview to present the login, and captures the redirect from Google to extract the code and call the authViaOAuth2 method.

**Comment by ganigeorgiev** [comment, marked as answer]
I haven't used webview_flutter, but as far as I know, Google doesn't allow webviews as context for their OAuth2 endpoints anymore (https://developers.googleblog.com/2021/06/upcoming-security-changes-to-googles-oauth-2.0-authorization-endpoint.html) and the recommendation is to start the device browser (something like https://pub.dev/packages/flutter_web_auth). 

The above example also doesn't seem to follow the PKCE flow (or at least I can't find where `code_challenge` and `code_challenge_method` are set).

Instead of manually constructing the google auth api url, there is an endpoint that will generate it for you by just calling
```dart
final authMethods = await client.users.listAuthMethods();
```

The above will return [`AuthMethodsList` DTO](https://pub.dev/documentation/pocketbase/latest/pocketbase/AuthMethodsList-class.html) and you can extract the google auth provider from the `AuthMethodsList.authProviders`.

**Full js OAuth2 integration example could be found in "Auth via OAuth2" section in https://pocketbase.io/docs/manage-users/#auth-via-oauth2. The flutter integration should be similar with the main difference that the redirect url will probably need to be a deep-link.**

_There are plans to simplify the OAuth2 flow in the future. You can subscribe to issue https://github.com/pocketbase/pocketbase/issues/55._

**Reply by pchasco** [reply]
Bummer about not being able to use a webview. That is certainly a better user experience than being taken out of the app. Obviously, no one should violate the terms of use, but I wonder how Google even can tell the difference between a request from a webview and one from a browser?

For now, I can use email auth to continue developing my application in anticipation of issue #55 resolution. If #55 is not yet ready, I can implement auth with the external browser and deep link.

Thank you!

**Reply by ganigeorgiev** [reply]
>  I wonder how Google even can tell the difference between a request from a webview and one from a browser?

Oh sure there is a way with some crafted "fingerprinting" to make the webview behave like a regular browser, but it is a lot safer for the users to see the auth page in the device browser (atleast they can verify that the Google auth page you are showing is the actual one and not some phishing login form).

_In addition, this seems to be also part of their "User data policy" https://developers.google.com/terms/api-services-user-data-policy#deceptive_or_unauthorized_use_of_google_api_services_is_prohibited which may lead to your app removal (and in the worst case scenario to legal actions against you)._

**Reply by pchasco** [reply]
It looks like Google does allow the use of Android Custom Tabs and SFSafariWebView for auth, according to the blog post (https://developers.googleblog.com/2021/06/upcoming-security-changes-to-googles-oauth-2.0-authorization-endpoint.html?m=1). I think these components will provide a reasonably good UX while complying with their terms. It also so happens that there appears to be a Flutter package that provides a common interface to the two platform components: https://pub.dev/packages/flutter_custom_tabs

**Reply by ganigeorgiev** [reply]
Yes, custom tabs are an option. The above linked https://pub.dev/packages/flutter_web_auth package is also using them (at least on Android).

**Comment by pchasco** [comment]
Returning to this after some time, and I am still unable to get Google auth to work in a Flutter app using the Dart SDK. I am not certain, but I wonder if the Google auth integration in its current form in PocketBase is unable to support Android applications. PocketBase requires both Client ID and Client Secret to configure Google auth, but Google does not provide a client secret for Android clients, so it is not possible to configure an Android client in the PocketBase interface. Google provides a client secret for web application clients, but it does not allow for use of a custom URL scheme for the redirect URL, which is required for the app to capture the redirect and complete the sign-in process.

**Reply by ganigeorgiev** [reply]
I'll have to investigate this more thoroughly but you should be able to use the "Web application" option to set a deep link and capture it in your android/flutter app.

If Google doesn't allow different url schemes, then you can try using a "dummy" url as a redirect address, eg. "https://myapp.com/android-redirect" and capture this url in your android/flutter app.

Checking the [documentation for the AndroidManifest](https://developer.android.com/guide/topics/manifest/data-element), you should be able to define something like:

```xml
<intent-filter ...>
    ...
    <data android:scheme="https" android:host="myapp.com" android:path="/android-redirect" />
</intent-filter>
```

**Reply by pchasco** [reply]
Thanks! It didn't occur to me to use an https scheme for the redirect, as the flutter_web_auth documentation indicates you should use a custom scheme.

With the changes you suggest, it <i>almost</i> works. The intent filter does not capture the redirect, but if I use the menu button in the Chrome Custom Tabs view after the redirect fails to load, there is an option to open the URL in Chrome. That actually triggers Android's app selector, and my app is one of the available options. Selecting my app does resume the login process and the authentication completes. I do not think that issue is in any way related to PocketBase, so I think PocketBase is working correctly.

**Reply by pchasco** [reply]
I have been able to get this working end-to-end. There is a peculiarity with Android that in order to support deep links with intent filter for an http or https scheme, you must be able to prove that you own the domain. This involves crafting a JSON document and hosting it at a defined URL on a public website that matches with the hostname in the intent filter. https://developer.android.com/training/app-links/verify-android-applinks

**Reply by alvinsga** [reply]
@pchasco could you post a full example of your working implementation? Really struggling as there is no documentation on how to implement this in Flutter.

**Reply by alvinsga** [reply]
> Thanks! It didn't occur to me to use an https scheme for the redirect, as the flutter_web_auth documentation indicates you should use a custom scheme.
> 
> With the changes you suggest, it almost works. The intent filter does not capture the redirect, but if I use the menu button in the Chrome Custom Tabs view after the redirect fails to load, there is an option to open the URL in Chrome. That actually triggers Android's app selector, and my app is one of the available options. Selecting my app does resume the login process and the authentication completes. I do not think that issue is in any way related to PocketBase, so I think PocketBase is working correctly.

How did you resolve the issue of having to open the URL in Chrome? I am having the same issue where the redirect is working fine but I need to open the website in Chrome for it to deep link back to the app and retrieve the URL.

**Comment by pchasco** [comment]
Sure. I’ll try to do so in the next day or two.

**Reply by lostintheway** [reply]
@pchasco  any updates?

**Reply by pchasco** [reply]
It's rushed, but I hope it helps.

https://github.com/pchasco/pocketbase_google_login_flutter

**Comment by pchasco** [comment]
Sorry. I’ll see if I can tonight. I’ve got 6-month old twins so my free
coding time is somewhat limited at the moment!

**Reply by pchasco** [reply]
https://github.com/pchasco/pocketbase_google_login_flutter

**Comment by Relorer** [comment]
I'm probably missing something, but I couldn't get the application to open https redirect.
So I added on the server index.html
```
<!DOCTYPE html>
<html>
<head>
  <title>Redirecting...</title>
</head>
<body>
  Redirecting...
  <script>
    const queryString = window.location.search;
    const urlParams = new URLSearchParams(queryString);
    const redirectUrl = `com.example.app://login-callback${queryString}`;
    window.location.replace(redirectUrl);
  </script>
</body>
</html>
```
All it does is redirect to com.example.app://login-callback keeping all parameters

Usage example
```
  const redirectUri = "https://yourdomen.com/index.html";
  const callbackUrlScheme = "com.example.app";
  
  final pb = PocketBase('https://yourdomen.com');
  
  final authMethods = await pb.collection("users").listAuthMethods();
  
  final google = authMethods.authProviders
      .where((am) => am.name.toLowerCase() == "google")
      .first;
  
  final responseUrl = await FlutterWebAuth.authenticate(
      url: "${google.authUrl}$redirectUri",
      callbackUrlScheme: callbackUrlScheme);
  
  final parsedUri = Uri.parse(responseUrl);
  final code = parsedUri.queryParameters['code']!;
  var result = await pb
      .collection("users")
      .authWithOAuth2("google", code, google.codeVerifier, redirectUri);
```

```
<activity android:name="com.linusu.flutter_web_auth.CallbackActivity" android:exported="true">
	<intent-filter android:label="flutter_web_auth">
		<action android:name="android.intent.action.VIEW" />
		<category android:name="android.intent.category.DEFAULT" />
		<category android:name="android.intent.category.BROWSABLE" />
		<data android:scheme="com.example.app" />
	</intent-filter>
</activity>
```

How it works
1. Start authorization
2. Redirect to index.html , accessed via https
3. Redirect to com.example.app://login-callback
4. The user has returned to the application, as com.example.app requests are processed by the application

I suppose something similar is built into supabase

**Reply by pchasco** [reply]
You do not need host a redirect anywhere. What you need to host is a file located at `https://yourdomen.com/.well-known/assetlinks.json` containing metadata that Google uses to validate that you own the app.

The file looks something like this:

```JSON
[
  {
    "relation": ["delegate_permission/common.handle_all_urls"],
    "target": {
      "namespace": "android_app",
      "package_name": "your app's complete package name here",
      "sha256_cert_fingerprints": [
        "sha fingerprint here"
      ]
    }
  }
]
```

You can find the full documentation here: https://developer.android.com/training/app-links/verify-android-applinks

**Reply by Relorer** [reply]
Thanks for the reply. I took some time to realize that the problem was the wrong signature.

Some useful information for those who will face a similar problem:
`adb shell pm get-app-links <PACKAGE_NAME>` - shows information about domain verification results (In particular, it outputs the signature of the apk file, which of course must match the signature in assetlinks.json)
https://developers.google.com/digital-asset-links/tools/generator - generating assetlinks.json

https://stackoverflow.com/questions/68144323/app-links-legacy-failure-verification-error-on-android-12
